### PR TITLE
koord-scheduler: add limit-aware scheduling plugin

### PR DESCRIPTION
**Pull request by ZiMengSheng**
### Ⅰ. Describe what this PR does
This PR implements a resource limit aware plugin to mitigate the risk of overcommitment.

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

**Review by eahydra — COMMENTED** [review]
This scheduling feature is quite important, and it also involves API changes. A specific proposal needs to be provided first.

File: pkg/scheduler/plugins/limitaware/pod_eventhandler.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+func assignedPod(pod *corev1.Pod) bool {
+	return pod.Spec.NodeName != ""
+}
+
+func (c *podEventHandler) OnAdd(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return
+	}
+	c.nodeLimitsCache.AddPod(pod.Spec.NodeName, pod)
```
Comment (eahydra) [inline pkg/scheduler/plugins/limitaware/pod_eventhandler.go:74]: Hmm,  the pod maybe terminated too, so should skip it?

**Comment by eahydra** [comment]
Please add more UT.

File: pkg/scheduler/plugins/limitaware/cache.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		return
+	}
+
+	nodeLimitsAllocated, ok := c.limitsAllocated[nodeName]
+	if !ok {
+		nodeLimitsAllocated = framework.NewResource(nil)
+		c.limitsAllocated[nodeName] = nodeLimitsAllocated
+	}
+	nodeLimitsAllocated.MilliCPU += podResourceLimit.MilliCPU
+	nodeLimitsAllocated.Memory += podResourceLimit.Memory
```
Comment (saintube) [inline pkg/scheduler/plugins/limitaware/cache.go:76]: EphemeralStorage is missing

File: apis/extension/node.go [hunk 3]
```diff
@@ -128,6 +130,28 @@ func GetNodeCPUSharePools(nodeTopoAnnotations map[string]string) ([]CPUSharedPoo
 	return cpuSharePools, nil
 }
 
+type LimitToAllocatableRatio map[corev1.ResourceName]float64
```
[the remark this answers is not in the document]
Comment (ZiMengSheng) [reply apis/extension/node.go:133]: Currently, we don't need the specific capacity of resources to be specified by user.

File: pkg/scheduler/apis/config/types.go [hunk 4]
```diff
@@ -208,3 +208,16 @@ type DeviceShareArgs struct {
...
+
+// LimitAwareArgs defines the parameters for LimitAware plugin.
+type LimitAwareArgs struct {
+	metav1.TypeMeta
+
+	// Resources a list of pairs <resource, weight> to be considered while scoring
+	// allowed weights start from 1.
+	Resources []schedconfig.ResourceSpec
+
+	DefaultLimitToAllocatableRatio extension.LimitToAllocatableRatio
```
[the remark this answers is not in the document]
Comment (ZiMengSheng) [reply pkg/scheduler/apis/config/types.go:222]: It will be used in both Filter and Score stage.

**Pull request closed without merging**